Add Mesh and Locstream to Geom

weiyuan-jiang opened the pull request:
## Types of change(s)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)
- [ ] Refactor (no functional changes, no api changes)

## Checklist
- [x] Tested this change with a run of GEOSgcm
- [x] Ran the Unit Tests (`make tests`)

## Description
for discussion
## Related Issue

tclune committed:
recreated reproducer for ESMF problem.

tclune committed:
Oops - did not commit files.

weiyuan-jiang committed:
re-order elements and vertices

mathomp4 commented:
Huh. Lots of CI failures. ninja seems unhappy with:
```
ninja: build stopped: multiple rules generate mesh2/sf_direction.mod.
```

As for ifx, not sure. Seems like our spurious random failure.

Also from GNU:
```
Error copying Fortran module "mesh2/sf_direction.mod".  Tried "mesh2/SF_DIRECTION.mod" and "mesh2/sf_direction.mod".
make[2]: *** [mesh2/CMakeFiles/to_mesh2.x.dir/depend.make:7: mesh2/CMakeFiles/to_mesh2.x.dir/sf_direction.mod.stamp] Error 1
make[1]: *** [CMakeFiles/Makefile2:9260: mesh2/CMakeFiles/to_mesh2.x.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 89%] Building Fortran object mesh2/CMakeFiles/xgrid.x.dir/Direction.F90.o
Error copying Fortran module "mesh2/sf_direction.mod".  Tried "mesh2/SF_DIRECTION.mod" and "mesh2/sf_direction.mod".
make[2]: *** [mesh2/CMakeFiles/xgrid2.x.dir/depend.make:7: mesh2/CMakeFiles/xgrid2.x.dir/sf_direction.mod.stamp] Error 1
make[1]: *** [CMakeFiles/Makefile2:9356: mesh2/CMakeFiles/xgrid2.x.dir/all] Error 2
```

weiyuan-jiang committed:
bug fixed and re-ordered

weiyuan-jiang commented:
There are a few make_esmf_mesh subroutines.  My tests show only the [one](https://github.com/GEOS-ESM/MAPL/blob/7d25f07ad9672b19becbfd6a209cf6dea3a345ed/mesh2/Mesh.F90#L46) works. @tclune,  We should clean those up

weiyuan-jiang committed:
remove wrong comments

weiyuan-jiang committed:
clean up. remove some subroutines

weiyuan-jiang committed:
added ordered multi connections outout and four connections

weiyuan-jiang committed:
error exposed

weiyuan-jiang committed:
refactoring...

tclune committed:
Add LocStream domain decomposition support

This commit implements MPI-parallel domain decomposition for LocStream
geometry types in MAPL, enabling efficient distributed processing of
location-based data (e.g., station observations, satellite swaths).

Key changes:

- Added LocStreamDecomposition class to manage 1D point distribution
  across MPI processes using mapl_GetPartition for even load balancing

- Integrated decomposition into LocStreamGeomSpec to store decomposition
  metadata alongside geometry specifications

- Updated LocStreamGeomFactory to apply decompositions when creating
  distributed LocStreams, extracting local coordinate subsets (i_0:i_1)
  based on each PE's assigned point range

- Comprehensive test suite with 6 tests covering:
  * Decomposition constructors and equality
  * Local index calculation across multiple PE counts
  * Actual LocStream creation with proper distribution on 1, 2, and 4 PETs
  * Verification of coordinate value cor  * Verification of coordinate value cor  * Verification of coordinate value cor  * Verificosi  * Verification of coordinate value cor  * Verification of coordinaubsystem.

Tested with NAG 7.2.41 compiler - all tests passing.

tclune reviewed (dismissed):
At some point we will ned to find a proper home and really clean this up.

weiyuan-jiang committed:
add command line arguments

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
mv mesh generating codes to geom/Mesh/utils

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/locstream-domain-decomposition

weiyuan-jiang committed:
Merge branch 'feature/locstream-domain-decomposition' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
initial check in Mesh ( copy from locstream)

weiyuan-jiang committed:
changed by claude-4.5-sonnet

weiyuan-jiang committed:
adding parallelization of mesh nodes

weiyuan-jiang committed:
fixed field_destroy whee there is no name and return RC

weiyuan-jiang committed:
change MeshGeom data flow and add more tests

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
merge and implement necessary interfaces

weiyuan-jiang committed:
fix hanging

weiyuan-jiang committed:
fix reading from netcdf4 for mesh

weiyuan-jiang committed:
comment out failed tests. will continue...

weiyuan-jiang committed:
Fix Phase 3 distributed mesh creation and re-enable 4PE test

This commit fixes multiple issues that caused the test_make_geom_4pe_four_quads
test to hang when creating distributed meshes with node ownership (Phase 3).

Issues Fixed:

1. Global-to-local node ID mapping (MeshGeomFactory_smod.F90)
   - ESMF requires element connectivity to use LOCAL node indices (positions
     in the nodeIds array), not global node IDs from the mesh file
   - Added global_to_local(:) array to map global node IDs to local indices
   - Convert element connectivity before passing to ESMF_MeshAddElements

2. Missing synchronization barriers (MeshGeomFactory_smod.F90)
   - Added ESMF_VMBarrier after ESMF_MeshAddNodes (collective operation)
   - Added ESMF_VMBarrier after ESMF_GeomCreate (collective operation)
   - These barriers ensure all PEs complete collective operations before
     proceeding to the next step

3. Restored elementMask functionality (MeshGeomFactory_smod.F90)
   - elementMask was previously disabled due to hanging issues
   - With the above fixes, elementMask now works correctly
   - Element mask values (Ocean/Land classification) are properly passed
     to ESMF_MeshAddElements

4. Test synchronization issues (Test_MeshGeomFactory_File.pf)
   - Added barriers before and after ESMF_GeomDestroy in test cleanup
   - Replaced _RC macro with explicit rc=status in cleanup_test_file call
     to avoid macro expansion issues in multi-PE context
   - Added final barrier before test exit to ensure all PEs finish together
   - Removed global count verification (ESMF_MeshGet returns LOCAL counts
     for distributed meshes)

5. Replaced _RC macros in test files (Test_MeshGeomFactory_File.pf, MeshTestHelper.F90)
   - Changed all _RC macros to explicit rc=status assignments
   - The _RC macro can cause unexpected behavior in test code, particularly
     in multi-PE scenarios where _VERIFY might trigger collective operations
   - Total replacements: 157 (48 in Test_MeshGeomFactory_File.pf,
     109 in MeshTestHelper.F90)

Test Results:
- test_make_geom_4pe_four_quads now passes successfully with 4 PEs
- All 7 MeshGeomFactory_File tests pass
- No regressions introduced

Files Modified:
- geom/Mesh/MeshGeomFactory_smod.F90
- geom/tests/Test_MeshGeomFactory_File.pf
- geom/tests/MeshTestHelper.F90

weiyuan-jiang committed:
Enable 4 additional 4PE distributed mesh tests

This commit enables 4 more multi-PE mesh tests that now pass with the
Phase 3 distributed mesh fixes. The tests verify mesh creation with
different element distributions and mixed element types.

Tests Enabled:
1. test_make_geom_4pe_eight_elements - 8 elements, 2 per PE
   - Tests even distribution with more elements than the basic test
   - Uses 5x3 node grid (15 nodes) with 4x2 element grid (8 quads)

2. test_make_geom_4pe_mixed_elements - Mixed triangles and quads
   - Tests Phase 3 mesh with different element types
   - 1 element per PE with varying node counts

3. test_make_geom_4pe_element_distribution - Verifies decomposition
   - Tests that element distribution indices match expectations
   - Verifies each PE gets correct element range (1-2, 3-4, 5-6, 7-8)

All tests updated with proper synchronization:
- Added ESMF_VMBarrier before ESMF_GeomDestroy
- Added ESMF_VMBarrier after ESMF_GeomDestroy
- Added ESMF_VMBarrier after cleanup_test_file
- Removed global count verification (ESMF_MeshGet returns local counts
  for distributed meshes)

Tests Remaining Disabled:
- test_make_geom_4pe_uneven_distribution: Uneven distribution [3,2,2,1]
  causes ESMF error 509. Requires additional node distribution logic.

- test_make_geom_4pe_from_hconfig: HConfig doesn't properly set up
  decomposition for multi-PE scenarios, causes ESMF error 509.

Test Results:
- Total MeshGeomFactory_File tests: 10 (up from 7)
- 6 single-PE tests: All passing
- 4 multi-PE (4PE) tests: All passing
- 2 tests remain disabled pending further investigation

Files Modified:
- geom/tests/Test_MeshGeomFactory_File.pf

weiyuan-jiang committed:
Fix HConfig multi-PE mesh decomposition

This commit fixes the test_make_geom_4pe_from_hconfig test by ensuring
that MeshGeomSpec created from HConfig/FileMetadata includes both element
and node distribution for Phase 3 distributed meshes.

Root Cause:
When creating a MeshGeomSpec from HConfig (which reads from FileMetadata),
the decomposition was created using only the element count:
  decomp = make_MeshDecomposition(nelements, _RC)

This creates a decomposition with only point_distribution (elements) but
not node_distribution, which is required for Phase 3 distributed meshes
where nodes are shared across PEs.

Solution:
Updated make_MeshGeomSpec_from_metadata to use the full constructor:
  decomp = MeshDecomposition(nelements, nnodes, petCount=petCount)

This ensures both element and node distributions are set up, enabling
proper Phase 3 mesh creation with node ownership.

Changes:
- geom/Mesh/MeshGeomSpec_smod.F90:
  * Get petCount from current VM
  * Use full MeshDecomposition constructor with both nelements and nnodes
  * This matches the pattern used in working multi-PE tests

- geom/tests/Test_MeshGeomFactory_File.pf:
  * Re-enabled test_make_geom_4pe_from_hconfig

Test Results:
- test_make_geom_4pe_from_hconfig now passes
- Total passing tests: 11 (6 single-PE + 5 multi-PE)
- 5 out of 6 4PE tests now pass
- Only test_make_geom_4pe_uneven_distribution remains disabled

Files Modified:
- geom/Mesh/MeshGeomSpec_smod.F90
- geom/tests/Test_MeshGeomFactory_File.pf

weiyuan-jiang committed:
Fix uneven element distribution with custom node distribution

This commit fixes test_make_geom_4pe_uneven_distribution by adding support
for specifying both element and node distributions when creating a
MeshDecomposition with custom (uneven) distributions.

Problem:
The test uses uneven element distribution [3,2,2,1] across 4 PEs (PE0 gets
3 elements, PE1 gets 2, PE2 gets 2, PE3 gets 1). When using the basic
MeshDecomposition constructor with an array, only point_distribution
(elements) was set, but not node_distribution. This caused Phase 3
distributed mesh creation to fail with ESMF error 509 because node
distribution is required for Phase 3.

Root Cause:
The basic constructor MeshDecomposition([3,2,2,1]) only sets element
distribution but Phase 3 requires both element AND node distribution
to properly handle shared nodes across PEs.

Solution:
1. Added new_MeshDecomposition_with_nodes constructor that accepts both
   element_distribution and node_distribution arrays as parameters

2. Updated the test to provide node distribution:
   MeshDecomposition([3,2,2,1], node_distribution=[4,4,4,3])

   For the 8-element, 15-node mesh:
   - Element distribution: uneven [3,2,2,1] (total=8)
   - Node distribution: even [4,4,4,3] (total=15)

   Even node distribution works fine with uneven element distribution
   because nodes are shared across element boundaries.

Changes:
- geom/Mesh/MeshDecomposition.F90:
  * Added new_MeshDecomposition_with_nodes constructor
  * Takes element_distribution and node_distribution arrays
  * Sets both point_distribution and node_distribution

- geom/tests/Test_MeshGeomFactory_File.pf:
  * Re-enabled test_make_geom_4pe_uneven_distribution
  * Added node_distribution=[4,4,4,3] to MeshDecomposition call
  * Added comment explaining the distribution strategy

Test Results:
- test_make_geom_4pe_uneven_distribution now passes
- All 12 MeshGeomFactory_File tests pass (6 single-PE + 6 multi-PE)
- All 6 4PE distributed mesh tests now pass

Files Modified:
- geom/Mesh/MeshDecomposition.F90
- geom/tests/Test_MeshGeomFactory_File.pf

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
merge release and change log

weiyuan-jiang committed:
remove unnecessary files

tclune reviewed (commented):
The `utils` directory seems poorly named.  (A bit vague.)

A bigger issue that is only occurring to me now.  We need to more carefully distinguish between the offline application that is used to generate a mesh (specific algorithm that we developed) from the generic mesh logic that interfaces MAPL's geom manager with ESMF.    Indeed it is possible that we will need multiple distinct factories for mesh just like we do for grid  (latlon, ease, tripolar, xy, cubed-sphere).  (Admittedly some of these could be merged by adding parameters but these distinctions are friendly to the user.)

I think there should be an app dir that produces the standalone driver for generating a mesh from a pixel map.   the program `xgrid.x` needs a better name.   And we want your benchmark program to go in here somehow as well.

We need to meet to discuss what here is generic and what is specific to the pixel generated mesh.

weiyuan-jiang committed:
Replace get_command_argument with fargparse

Both gen_mesh.x and xgrid.x now use fargparse for argument
parsing instead of manual get_command_argument loops.

Changes:
- Add fargparse import to both programs
- Create parse_command_line() functions using ArgParser
- Support short and long options (-r/--resolution, -o/--output, etc.)
- Use allocatable strings for compatibility with fargparse cast()
- Add FARGPARSE::fargparse linkage in CMakeLists.txt
- Remove manual print_usage() subroutines (fargparse auto-generates help)

Benefits:
- Cleaner, more maintainable code
- Automatic --help generation
- Consistent with other MAPL utilities (checkpoint_simulator, etc.)
- Better type safety and error handling

weiyuan-jiang committed:
remove debug print

weiyuan-jiang committed:
Update geom/Mesh/utils/gen_mesh.F90

Co-authored-by: Tom Clune <thomas.l.clune@nasa.gov>

weiyuan-jiang commented on geom/Mesh/utils/Direction.F90:
```diff
@@ -0,0 +1,48 @@
+   public :: next_direction
+   public :: dir_string
+
+   enum, bind(c)
```
Added

weiyuan-jiang committed:
added critical comment for mesh element direction

weiyuan-jiang committed:
remove obsolete lines

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
refactoring... change names

weiyuan-jiang committed:
Merge branch 'feature/wjiang/task2_reorder' of github.com:GEOS-ESM/MAPL into feature/wjiang/task2_reorder

weiyuan-jiang committed:
Optimize mesh creation from files with automatic and custom distribution paths

This commit refactors MeshGeomFactory to use ESMF_MeshCreate directly when
reading mesh files, providing significant performance improvements through
ESMF's native parallel I/O while maintaining support for custom decompositions.

Changes:
- Add is_uniform_decomposition() to detect default even element distributions
- Add decomp_to_distgrid() to convert custom MeshDecomposition to ESMF_DistGrid
- Modify create_esmf_mesh() with two optimized paths:
  * Path 1: Automatic distribution for uniform decompositions (most common case)
  * Path 2: Custom distribution via elementDistgrid for non-uniform cases
- Remove manual file reading when using ESMF's built-in reader
- Maintain backward compatibility with programmatically-created meshes

Benefits:
- Simplified code: removes ~50 lines of manual file reading for uniform cases
- Better performance: leverages ESMF's optimized parallel I/O
- Flexibility: supports both automatic and custom distributions transparently
- No API changes: decision made automatically based on decomposition type

Tests:
- Add comprehensive test suite (Test_MeshGeomFactory_Distribution.pf) with 5 tests
- All tests pass: automatic distribution, custom distribution, remainder handling,
  single-PE case, and backward compatibility
- No existing tests broken

weiyuan-jiang committed:
Remove pointer-based mesh data and convert to file-based mesh creation

This commit refactors MeshGeomSpec to remove in-memory pointer storage
for mesh data (node_coords, element_conn, num_element_conn, element_mask)
and simplifies mesh creation to use ESMF's file-based path exclusively.

Changes:
- Remove pointer data members and related methods (set_mesh_data, get_node_coords,
  get_connectivity, get_num_element_conn, get_element_mask) from MeshGeomSpec
- Simplify create_esmf_mesh to only create meshes from ESMF-format files using
  ESMF_MeshCreate, supporting both automatic and custom elementDistgrid distribution
- Reduce MeshGeomFactory_smod.F90 create_esmf_mesh from 291 lines to 42 lines
- Update typesafe_make_file_metadata to generate generic metadata without requiring
  connectivity data access
- Add helper functions for hexagon and octagon mesh file creation
- Convert all tests to use file-based mesh creation via MeshTestHelper functions
- Remove backward compatibility tests for programmatic mesh creation
- Move _UNUSED_DUMMY declarations to end of subroutines (before _RETURN)

Test results: All mesh-related tests pass (59/59 single-PE, 15/15 multi-PE)

Net result: -328 lines of code, cleaner architecture, better performance

weiyuan-jiang committed:
Replace MAPL_ErrLog.h with MAPL.h in geom/Mesh folder

Update include directive from MAPL_ErrLog.h to MAPL.h in all Fortran
files within geom/Mesh and its subfolders for consistency with MAPL
header conventions.

Files updated:
- MeshDecomposition.F90
- MeshDecomposition_smod.F90
- MeshGeomFactory.F90
- MeshGeomFactory_smod.F90
- MeshGeomSpec.F90
- MeshGeomSpec_smod.F90
- raster_to_mesh/MeshVertex.F90
- raster_to_mesh/raster_to_mesh.F90

weiyuan-jiang committed:
Remove nnodes/nelements from MeshGeomSpec

Remove nnodes and nelements as stored member variables and
constructor arguments. The constructor now only accepts filename
and optional decomposition. Accessor methods get_nnodes() and
get_nelements() now read dimensions directly from the mesh file
on-demand using pfio's NetCDF4_FileFormatter.

This simplifies the API and ensures consistency with file contents.

weiyuan-jiang committed:
Simplify make_file_metadata to store filename

Simplify typesafe_make_file_metadata function to only store the
mesh filename using set_source_file(), since the file contains
all mesh information (dimensions, coordinates, connectivity).

Update supports_metadata to check if FileMetadata has source_file
set and verify the file contains valid mesh structure by reading it.

This eliminates data duplication and reduces the function from
~35 lines to ~10 lines.

weiyuan-jiang committed:
Update tests for file-based MeshGeomSpec

Update all test files to use the new file-based constructor:
- Create mesh files using MeshTestHelper functions
- Pass filename to MeshGeomSpec constructor
- Clean up test files after use

Fix test_filename_accessor to create file before testing.
Update test_make_file_metadata_from_spec to verify source_file.

All 40 mesh-related tests pass with this update.

tclune committed:
Apply suggestion from @tclune

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
change CMakeLists.txt

weiyuan-jiang committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

weiyuan-jiang committed:
Don't build reaster_to_mesh if DBUILD_WITH_FARGPARSE is off

weiyuan-jiang committed:
Merge branch 'feature/wjiang/task2_reorder' of github.com:GEOS-ESM/MAPL into feature/wjiang/task2_reorder

mathomp4 committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

mathomp4 committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

tclune committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder

mathomp4 committed:
Merge branch 'release/MAPL-v3' into feature/wjiang/task2_reorder
